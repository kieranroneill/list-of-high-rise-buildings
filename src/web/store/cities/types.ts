import { Action } from 'redux';

// Types.
import {
    ApiErrorAction,
    ApiRequestAction,
    ApiSuccessAction,
} from '../api/types';

//====================================================
// Action types.
//====================================================

export enum CitiesActionTypes {
    GetCities = '@cities/GET_CITIES',
    GetCitiesError = '@cities/GET_CITIES_ERROR',
    GetCitiesRequest = '@cities/GET_CITIES_REQUEST',
    GetCitiesSuccess = '@cities/GET_CITIES_SUCCESS',
}

//====================================================
// Actions.
//====================================================

export interface GetCitiesAction extends Action {
    type: CitiesActionTypes.GetCities;
}

export type GetCitiesErrorAction = ApiErrorAction<CitiesActionTypes.GetCitiesError>;

export type GetCitiesRequestAction = ApiRequestAction<CitiesActionTypes.GetCitiesRequest>;

export type GetCitiesSuccessAction = ApiSuccessAction<CitiesActionTypes.GetCitiesSuccess, City[]>;

export type CitiesActions = GetCitiesAction
    | GetCitiesErrorAction
    | GetCitiesRequestAction
    | GetCitiesSuccessAction;

//====================================================
// States.
//====================================================

export interface CitiesState {
    error: string,
    loading: boolean;
    cities: City[];
}
