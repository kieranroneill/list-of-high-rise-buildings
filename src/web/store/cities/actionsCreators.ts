// Action creators.
import { callApi } from '../api/actionCreators';

// Types.
import { CallApiAction } from '../api/types';
import { CitiesActionTypes } from './types';

export const getCities: () => CallApiAction = () => {
    return callApi(
        [
            CitiesActionTypes.GetCitiesRequest,
            CitiesActionTypes.GetCitiesSuccess,
            CitiesActionTypes.GetCitiesError,
        ],
        '/city',
        'GET'
    );
};
