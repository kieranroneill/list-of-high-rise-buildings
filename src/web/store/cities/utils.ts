// Types.
import { CitiesState } from './types';

export function getInitialState(): CitiesState {
    return {
        cities: [],
        errors: [],
        loading: false,
    };
}
