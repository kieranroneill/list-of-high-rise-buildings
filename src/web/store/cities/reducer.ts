import { Reducer } from 'redux';

// Types.
import {
    CitiesActions,
    CitiesActionTypes,
    CitiesState
} from './types';

// Utils.
import { getInitialState } from './utils';

const reducer: Reducer<CitiesState, CitiesActions> = (state: CitiesState = getInitialState(), action: CitiesActions) => {
    switch (action.type) {
        case CitiesActionTypes.GetCitiesError:
            return {
                ...state,
                errors: action.payload.errors,
                loading: false,
            };
        case CitiesActionTypes.GetCitiesRequest:
            return {
                ...state,
                loading: true,
            };
        case CitiesActionTypes.GetCitiesSuccess:
            return {
                ...state,
                cities: action.payload,
                loading: false,
            };
        default:
            return state;
    }
};

export default reducer;
