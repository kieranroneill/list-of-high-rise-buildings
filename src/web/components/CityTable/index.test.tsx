import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';

// Components.
import { CityTable, Props } from './';

// Mocks.
import { MockLocation } from '../../../../test/mocks/reactRouterMock';

// Store.
import { getInitialState as getInitialCitiesState } from '../../store/cities/utils';

interface Scope {
    props: Props;
    wrapper: ShallowWrapper;
}

describe('src/web/components/CityTable', () => {
    let scope: Scope;

    beforeEach(() => {
        const props: Props = {
            citiesState: getInitialCitiesState(),
            location: new MockLocation(),
        };

        scope = {
            props,
            wrapper: shallow(<CityTable {...props} />),
        };
    });

    describe('<CityTable /> snapshots', () => {
        it('should match the snapshot with the default props', () => {
            expect(scope.wrapper).toMatchSnapshot();
        });

        it('should match the snapshot when the cities is loading', () => {
            scope.wrapper.setProps({
                citiesState: {
                    ...scope.props.citiesState,
                    loading: true,
                },
            } as Partial<Props>);

            expect(scope.wrapper).toMatchSnapshot();
        });

        it('should display the table with values', () => {
            scope.wrapper.setProps({
                citiesState: {
                    ...scope.props.citiesState,
                    cities: [
                        {
                            '100m+': 87,
                            '150m+': 21,
                            '200m+': 3,
                            '300m+': 0,
                            allBuildings: 111,
                            allStructures: 114,
                            country: 'United States',
                            id: 44,
                            name: 'San Francisco',
                            telecomTowers: 3
                        }
                    ],
                    loading: false,
                },
            } as Partial<Props>);

            expect(scope.wrapper).toMatchSnapshot();
        });
    });
});
