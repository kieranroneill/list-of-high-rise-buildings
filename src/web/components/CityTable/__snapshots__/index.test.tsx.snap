// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/web/components/CityTable <CityTable /> snapshots should display the table with values 1`] = `
<styled.div>
  <WithStyles(Paper)>
    <WithStyles(Table)>
      <WithStyles(TableHead)>
        <WithStyles(TableRow)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              #
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              City
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              Country
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              All Buildings
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              100m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              150m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              200m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              300m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableHead)>
      <WithStyles(TableBody)>
        <WithStyles(TableRow)
          key="san-francisco"
        >
          <WithStyles(TableCell)
            component="th"
            scope="row"
          >
            44
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            San Francisco
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            United States
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            111
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            87
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            21
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            3
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </WithStyles(Paper)>
</styled.div>
`;

exports[`src/web/components/CityTable <CityTable /> snapshots should match the snapshot when the cities is loading 1`] = `
<styled.div>
  <WithStyles(Paper)>
    <WithStyles(CircularProgress)
      size={50}
    />
  </WithStyles(Paper)>
</styled.div>
`;

exports[`src/web/components/CityTable <CityTable /> snapshots should match the snapshot with the default props 1`] = `
<styled.div>
  <WithStyles(Paper)>
    <WithStyles(Table)>
      <WithStyles(TableHead)>
        <WithStyles(TableRow)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              #
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              City
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              Country
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              All Buildings
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              100m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              150m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              200m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
            >
              300m+
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableHead)>
      <WithStyles(TableBody) />
    </WithStyles(Table)>
  </WithStyles(Paper)>
</styled.div>
`;
